Adopt smart pointers in WebKit/UIProcess/BrowsingContextGroup.cpp

fred-wang opened the pull request:
#### 5853b8960b6d05972aef970edf40c4fc26f1d2a0
<pre>
Adopt smart pointers in WebKit/UIProcess/BrowsingContextGroup.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=286810">https://bugs.webkit.org/show_bug.cgi?id=286810</a>

Reviewed by Chris Dumez.

* Source/WebKit/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations: Remove expectation.
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations: Ditto.
* Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
(WebKit::BrowsingContextGroup::ensureProcessForConnection): FrameProcess::create() is non trivial, use protected version of WebPageProxy::legacyMainFrameProcess() as the WebProcessProxy argument.
(WebKit::BrowsingContextGroup::addFrameProcess): RemotePageProxy::create() is non trivial, hold a smart pointer to WebPageProxy and WebProcessProxy arguments.
(WebKit::BrowsingContextGroup::addPage): RemotePageProxy::create() is non trivial, hold a smart pointer to WebProcessProxy argument.
(WebKit::BrowsingContextGroup::takeRemotePageInProcessForProvisionalPage): HashSet::take() is non trivial, so hold a smart pointer to remotePage.
(WebKit::BrowsingContextGroup::transitionPageToRemotePage): RemotePageProxy::create() is non trivial, use protected version of WebPageProxy::legacyMainFrameProcess() as the WebProcessProxy argument.
(WebKit::BrowsingContextGroup::transitionProvisionalPageToRemotePage):
WeakHashMap::ensure is not trivial, use protected version of ProvisionalPageProxy::page().
RemotePageProxy::create() is non trivial, use protected version of WebPageProxy::process() as the WebProcessProxy argument.

Canonical link: <a href="https://commits.webkit.org/289735@main">https://commits.webkit.org/289735@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4f3bd5a763d9e1771b2999ce45ae17c9eebd686a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87798 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7314 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42185 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92663 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38548 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7695 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15485 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67791 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25534 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90800 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5882 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79442 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48159 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5669 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37655 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76067 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34729 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94550 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14966 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11005 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76638 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15221 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75298 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75874 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18661 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20242 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18682 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7968 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14982 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/20285 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14726 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18170 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16508 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fred-wang commented on Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
```diff
@@ -211,11 +213,11 @@ void BrowsingContextGroup::transitionPageToRemotePage(WebPageProxy& page, const
 void BrowsingContextGroup::transitionProvisionalPageToRemotePage(ProvisionalPageProxy& page, const Site& provisionalNavigationFailureSite)
 {
-    auto& set = m_remotePages.ensure(*page.page(), [] {
+    auto& set = m_remotePages.ensure(*page.protectedPage(), [] {
```
I'm a bit unsure for this one, it seems it's only used as a key for the hash map... how do we handle that normally?

cdumez replied:
cc @rniwa for this one. It does seem unnecessary so maybe we can improve static analysis here. That said, this is not hot code so I don't mind the extra ref'ing here.

fred-wang commented on Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
```diff
@@ -187,10 +189,10 @@ RefPtr<RemotePageProxy> BrowsingContextGroup::takeRemotePageInProcessForProvisio
     if (!remotePage)
         return nullptr;
-    return it->value.take(remotePage);
+    return it->value.take(remotePage.get());
```
I'm a bit unsure for this one, it seems it's only used as a key for the hash set... how do we handle that normally?

fred-wang replied:
I guess this is the same answer as below for BrowsingContextGroup::transitionProvisionalPageToRemotePage.

cdumez commented on Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
```diff
@@ -88,13 +88,14 @@ void BrowsingContextGroup::addFrameProcess(FrameProcess& process)
             return HashSet<Ref<RemotePageProxy>> { };
         }).iterator->value;
-        Ref newRemotePage = RemotePageProxy::create(page, process.process(), site);
+        Ref processProxy = process.process();
```
Ideally we'd move this before the loop to avoid refing/derefing the same process for each loop iteration.

fred-wang replied:
Good point. Done.

fred-wang committed:
Adopt smart pointers in WebKit/UIProcess/BrowsingContextGroup.cpp
https://bugs.webkit.org/show_bug.cgi?id=286810

Reviewed by Chris Dumez.

* Source/WebKit/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations: Remove expectation.
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations: Ditto.
* Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
(WebKit::BrowsingContextGroup::ensureProcessForConnection): FrameProcess::create() is non trivial, use protected version of WebPageProxy::legacyMainFrameProcess() as the WebProcessProxy argument.
(WebKit::BrowsingContextGroup::addFrameProcess): RemotePageProxy::create() is non trivial, hold a smart pointer to WebPageProxy and WebProcessProxy arguments.
(WebKit::BrowsingContextGroup::addPage): RemotePageProxy::create() is non trivial, hold a smart pointer to WebProcessProxy argument.
(WebKit::BrowsingContextGroup::takeRemotePageInProcessForProvisionalPage): HashSet::take() is non trivial, so hold a smart pointer to remotePage.
(WebKit::BrowsingContextGroup::transitionPageToRemotePage): RemotePageProxy::create() is non trivial, use protected version of WebPageProxy::legacyMainFrameProcess() as the WebProcessProxy argument.
(WebKit::BrowsingContextGroup::transitionProvisionalPageToRemotePage):
WeakHashMap::ensure is not trivial, use protected version of ProvisionalPageProxy::page().
RemotePageProxy::create() is non trivial, use protected version of WebPageProxy::process() as the WebProcessProxy argument.

Canonical link: https://commits.webkit.org/289735@main
